### ISSUE TITLE
DllReferencePlugin with multiple dll requires loading all

### ISSUE THREAD

**Issue by Jojoshua** [webpack/webpack#8078]
**What is the current behavior?**

I have 2 named items in my config to create dll bundles. The first one is intended as the base references for the whole site and I am including that on my main layout.html page. The 2nd one is only meant for a separate page and should be dynamically loaded by webpack. The issue is that I get an error when loading layout.html stating ....Uncaught ReferenceError: [the 2nd dll name] is not defined

I believe this is what issue https://github.com/webpack/webpack/issues/6656 was trying to say as well.

**If the current behavior is a bug, please provide the steps to reproduce.**

Here is an example entry for the dll config
`entry: {   
        base: [
            'bootstrap',            
        ],        
        jqueryui: ['jquery-ui/ui/widgets/draggable']
},   `

and loaded into the main configuration like so...
        new webpack.DllReferencePlugin({
            context: __dirname,
            manifest: require('./dist/dll/base.json')
        }),

        new webpack.DllReferencePlugin({
            context: process.cwd(),
            manifest: require('./dist/dll/jqueryui.json')
        }),

layout.html includes...
 <script src="/dist/dll/base.dll.js"></script>

**What is the expected behavior?**

loading layout.html does not require jqueryui.dll to be loaded as well. This is wasteful as it is not needed on this page. If I do load jqueryui.dll in layout.html it does not throw the error.

Dynamically load the dll into the page that it requires only. Do not throw an error if all dlls are not included.

**Other relevant information:**
webpack version:  4.18.1
Node.js version: 8.12.0
Operating System: Win 10
Additional tools:

**Comment by sokra** [comment]
Why do you reference the DLL when it's not needed?

**Comment by sokra** [comment]
Something in the module graph seems to depend on the DLL. If nothing depends on the DLL it would not ask for it at runtime.

You need a separate entrypoint where nothing references the DLL.

**Comment by Jojoshua** [comment]
Let me see if I understand what you mean. 

Ok so let's say jquery is defined in dll1 and jqueryui is defined in dll2. Jqueryui depends on jquery so I need to also include jquery in dll2 as well?

**Comment by sokra** [comment]
That's not what I wanted to say, but for this scenario you might need a DllReferencePlugin pointing to dll1 in the build of dll2.

**Comment by Jojoshua** [comment]
Is the dll plugin meant to be used with multiple entries? I haven't found any examples of it being used this way.

**Comment by willowTong** [comment]
Is this problem fixed?

**Comment by Vanuan** [comment]
@Jojoshua 
I suspect your layout html contains not only "base.dll.js", but also entrypoint of the file using dlls.

So it seemes to me you hoped that your entrypoint that references both dlls, but only uses modules from base.dll would not complain about missing modules from the second dll, jqueryui. Your expectation is based on the fact that you're not importing any modules from jqueryui from your entrypoint.

Well, your expectation is wrong. Webpack doesn't look into dll contents to determine whether your main entrypoint needs dll or not. Webpack is dumb and does only what you tell it to do.

In this case you told webpack that your main entrypoint references both manifests. So webpack believes you and includes both dll references into your main entrypoint.

Webpack doesn't optimize your dlls either. It includes whatever is required by dll's entrypoint. So if you included jqueryui (which depends on jquery) into jqueryui dll's entrypoint and included jquery into your base dll entrypoint, you must expect that jquery would be included twice.

That's one of the many issues with dll plugin. That's why it's only recommended in development. It's included tons and tons of unnecessary and duplicated stuff. It saves you some CPU cycles but spends a lot of space.

Unless you use multiple entrypoints (layout.js and some other entrypoint you want to load dynamically which isn't possible at the time) each referencing its own dll manifest, you can't expect to nave no errors in console.

@sokra I think Jojoshua means he has an error in his webpack main entrypoint. I suspect full webpack config looks similar to this:

```
// webpack.dll.config.js
entry: {
  base: [ 'bootstrap', ],
  jqueryui: ['jquery-ui/ui/widgets/draggable']
},
plugins: [
    new webpack.DllPlugin({
        path: './dist/dll/base.json',
        name:  '[name]_[dllHash]',
        context: process.cwd(),,
    })

    new webpack.DllPlugin({
        path: './dist/dll/jqueryui.json',
        name:  '[name]_[dllHash]',
        context: process.cwd(),,
    })

]
output = {
    filename: `[name].js`,
    library: '[name]_[dllHash]'
};


// webpack.main.config.js (to what Jojoshua referred as "main configuration")
entry: {
  layout: [ './layout.js', ],
  dynamicallyLoaded: ['./heavy.js']
},
plugins: [
    new webpack.DllReferencePlugin({
       context: __dirname,
       manifest: require('./dist/dll/base.json')
    }),

    new webpack.DllReferencePlugin({
        context: process.cwd(),
        manifest: require('./dist/dll/jqueryui.json')
    }),

]
output = {
    filename: `[name].js`,
    library: '[name]_[dllHash]'
};
```

And I also suspect those webpack configs are dynamically generated by some of the numerous project generators.

@Jojoshua Do your webpack files look similar to what I posted? Are you aware that all your entrypoints reference both dlls and webpack couldn't possibly know that your layout file only uses packages from one dll? If your "main configuration" doesn't actually reference both entrypoints and you have multiple configurations in a single "main configuration" file please clarify how do you include DllReferencePlugin into your main configuration.

**Comment by Vanuan** [comment]
@willowTong I suspect the issue comes from misunderstanding of how dll plugin works. This misunderstanding has three causes:

1. Bad documentation and complex configuration.
2. First cause resulted in various bootstrappers, helpers and **code generation scripts** that hide how webpack works. People expect webpack to "just work" and good enough. When they come across the issue they don't ask for help from code generation script maintainer, they ask for help from webpack. Since code generation script maintainers don't support code after it's generated.
3. Developers face enormous codebase generated by code generation script and just don't want to learn it and understand how it works. This leads to them not demanding good documentation. Since most of the job is done by code generator maintainer which is a hacker that doesn't need docs, he reads code directly. Lack of demand for documentation leads to square 1: webpack maintainer don't feel the pressure to write better docs since ordinary developers don't read webpack docs, they just copy something from blog posts.
4. When coming back to webpack developers can't even describe the issue they're having. Since they don't know which parts of config are generated by code generator (or copy pasted) and which they've written themselves by reading documentation and experimenting. So they don't know how much code needs to be posted to reproduce the issue. And they can't write steps to reproduce since they don't know how to create webpack config from scratch in the first place, since it was generated or copy pasted.

Some suggestions:

1. Webpack mainainers should write (or produce links to) blog posts about best practices. Each webpack feature should deserve a separate official blog post. Documentation tends to be dry and boring. There should be smiley faces and funny pictures. Millennials don't like to read :( .
2. It's best if webpack "just works" ©️ . E.g. `webpack start app.html` should work without configuration. This way developers can create reasonable amount of lines to demonstrate reproduction steps.
3. There should be configuration presets for various usage scenarios so that developers don't fiddle around the configs they don't understand. Babel is good at this, webpack could follow the lead.
4. If developers still don't know what's going on, webpack should print "effective" configuration file. And steps how to produce it should be provided in GitHub issue template. For sensitive developers there should be an option to obfuscate entrypoint names.


Well, probably I should've followed my own advice and write less text 🤷‍♂️ 
---
TL;DR You can't fix the issue. It's a feature, not a bug.

**Comment by Jojoshua** [comment]
@Vanuan 
It has been some time since I posted this issue and I changed the way I did things to resolve/workaround my issue. Where is it recommended that DLL plugin only be used in development?

My config used to looked something like...
// webpack.dll.config.js
entry: {
  base: [ 'bootstrap', ],
  jqueryui: ['jquery-ui/ui/widgets/draggable']
},

plugins: [  
        new webpack.DllPlugin({
            name: '[name]',
            path: 'dist/dll/[name].json'
        }),      
]

    output: {
        filename: '[name].dll.js',
        path: path.resolve(__dirname, "dist/dll"),
        library: '[name]'
    },

//webpack.main.config.js
plugins: [
    new webpack.DllReferencePlugin({
       context: __dirname,
       manifest: require('./dist/dll/base.json')
    }), 
]


The original question was if the DLL plugin worked having multiple entry points. I only had one reference to dllplugin for both entry points where the config you referenced earlier had 2. Maybe that would have fixed my issue, but I don't have time to try it. 

What I was trying to do what create multiple sets of DLLs within one config and then only reference the DLL entry I needed statically in the webpage but also make sure that my main webpack config entries knew that those DLLs existed and not to try and load them again and duplicate javascript. 

My config was not auto generated, it was hand crafted over a long period of time.

**Comment by Vanuan** [comment]
Ok, so much assumptions we're all wrong. Your newly posted configs look good to me. Are you saying you have reference error with it?

Also I didn't quite get what you meant by "if DllPlugin worked having multiple entrypoints". In webpack.dll config you can have multiple entrypoints, which would mean you have multiple dll js and multiple dll manifest files, completely independent of each other. Similarly to how you have entrypoints in your main config file. But having multiple entrypoints in a single config means parts of configuration are shared. You could use templating and name dlls similarly to main entrypoints. So that each main entrypoint would be mapped to corresponding dll entrypoint.

DllPlugin is naturally only for development since it's needed to reduce initial build time. For production it doesn't make sense to use it since your CI would build it from scratch anyway. And there's no advantage of dll plugin for initial build time. In fact, there's a performance overhead.

Having said that I'm now realizing you could have multiple build jobs one for each entrypoint. And setup some server to host manifest files which would be fetched during build. That would be like npm server for prebuild dlls. But in that case why not just use CDN? Don't see any advantage of using DllPlugin for production.
